app/scripts/ui.js

```diff
@@ -36,8 +40,36 @@ const WORKER_KEEP_ALIVE_MESSAGE = 'WORKER_KEEP_ALIVE_MESSAGE';
  * Time has been kept to 1000ms but can be reduced for even faster re-activation of service worker
  */
 if (isManifestV3) {
-  setInterval(() => {
+  const handle = setInterval(() => {
+    const messageId = new Date().getTime() + Math.random();
+    ACK_KEEP_ALIVE_RECEIVED[messageId] = false;
     browser.runtime.sendMessage({ name: WORKER_KEEP_ALIVE_MESSAGE });
+
+    /**
+     * set timeout to clear the ACK_KEEP_ALIVE_RECEIVED or
+     * show error if response is not received after ACK_KEEP_ALIVE_WAIT_TIME seconds
+     */
+    const timeoutHandle = setTimeout(() => {
+      if (ACK_KEEP_ALIVE_RECEIVED[messageId]) {
+        delete ACK_KEEP_ALIVE_RECEIVED[messageId];
+      } else {
+        clearInterval(handle);
+        displayCriticalError(
+          'somethingIsWrong',
+          new Error("Something's gone wrong. Try reloading the page."),
+        );
+      }
+    }, ACK_KEEP_ALIVE_WAIT_TIME);
```

Comment:
I am not clear about how is it useful to keep track of `messageId` here.

Reply:
I added it because we are sending the message every second while the interval to check timeout is a minute. It means by the time we get the interval, we need to check the particular message (identified with messageId) so that we don't show error for a message that has not been sent for less than a minute

Reply:
I see, just curious can we not keep a timestamp of when was last message received and see if it exceeds 1 minute.

Reply:
yeah, we could do that too.

Reply:
That will be better approach I think